File: apimsdc/SessMsdc.cpp
/**
  * \file SessMsdc.cpp
  * API code for job SessMsdc (implementation)
  * \author Alexander Wirthmueller
  * \date created: 29 Aug 2018
  * \date modified: 29 Aug 2018
  */

#ifdef _WIN32
	#include "stdafx.h"
#endif

#include "SessMsdc.h"

/******************************************************************************
 class SessMsdc::StatShr
 ******************************************************************************/

SessMsdc::StatShr::StatShr(
			const string& scrJrefCrdnav
		) :
			Block()
		{
	this->scrJrefCrdnav = scrJrefCrdnav;

	mask = {SCRJREFCRDNAV};
};

bool SessMsdc::StatShr::readXML(
			xmlXPathContext* docctx
			, string basexpath
			, bool addbasetag
		) {
	clear();

	bool basefound;

	if (addbasetag)
		basefound = checkUclcXPaths(docctx, basexpath, basexpath, "StatShrSessMsdc");
	else
		basefound = checkXPath(docctx, basexpath);

	string itemtag = "StatitemShrSessMsdc";

	if (basefound) {
		if (extractStringAttrUclc(docctx, basexpath, itemtag, "Si", "sref", "scrJrefCrdnav", scrJrefCrdnav)) add(SCRJREFCRDNAV);
	};

	return basefound;
};

set<uint> SessMsdc::StatShr::comm(
			const StatShr* comp
		) {
	set<uint> items;

	if (scrJrefCrdnav == comp->scrJrefCrdnav) insert(items, SCRJREFCRDNAV);

	return(items);
};

set<uint> SessMsdc::StatShr::diff(
			const StatShr* comp
		) {
	set<uint> commitems;
	set<uint> diffitems;

	commitems = comm(comp);

	diffitems = {SCRJREFCRDNAV};
	for (auto it=commitems.begin();it!=commitems.end();it++) diffitems.erase(*it);

	return(diffitems);
};

/******************************************************************************
 class SessMsdc::DpchEngData
 ******************************************************************************/

SessMsdc::DpchEngData::DpchEngData() :
			DpchEngMsdc(VecMsdcVDpch::DPCHENGSESSMSDCDATA)
		{
	feedFEnsSec.tag = "FeedFEnsSec";
};

string SessMsdc::DpchEngData::getSrefsMask() {
	vector<string> ss;
	string srefs;

	if (has(SCRJREF)) ss.push_back("scrJref");
	if (has(FEEDFENSSEC)) ss.push_back("feedFEnsSec");
	if (has(STATSHR)) ss.push_back("statshr");

	StrMod::vectorToString(ss, srefs);

	return(srefs);
};

void SessMsdc::DpchEngData::readXML(
			xmlXPathContext* docctx
			, string basexpath
			, bool addbasetag
		) {
	clear();

	bool basefound;

	if (addbasetag)
		basefound = checkUclcXPaths(docctx, basexpath, basexpath, "DpchEngSessMsdcData");
	else
		basefound = checkXPath(docctx, basexpath);

	if (basefound) {
		if (extractStringUclc(docctx, basexpath, "scrJref", "", scrJref)) add(SCRJREF);
		if (feedFEnsSec.readXML(docctx, basexpath, true)) add(FEEDFENSSEC);
		if (statshr.readXML(docctx, basexpath, true)) add(STATSHR);
	} else {
		feedFEnsSec.clear();
		statshr = StatShr();
	};
};

File: japimsdc/SessMsdc.java
/**
  * \file SessMsdc.java
  * Java API code for job SessMsdc
  * \author Alexander Wirthmueller
  * \date created: 18 Dec 2018
  * \date modified: 18 Dec 2018
  */

package apimsdc;

import java.util.*;
import org.w3c.dom.*;
import sbecore.*;

public class SessMsdc {
	/**
	  * StatShr (full: StatShrSessMsdc)
	  */
	public class StatShr extends Block {

		public static final int SCRJREFCRDNAV = 1;

		public StatShr(
					String scrJrefCrdnav
				) {
			this.scrJrefCrdnav = scrJrefCrdnav;

			mask = new HashSet<Integer>(Arrays.asList(SCRJREFCRDNAV));
		};

		public String scrJrefCrdnav;

		public boolean readXML(
					Document doc
					, String basexpath
					, boolean addbasetag
				) {

			clear();

			if (addbasetag) basexpath = Xmlio.checkUclcXPaths(doc, basexpath, "StatShrSessMsdc");

			String itemtag = "StatitemShrSessMsdc";

			if (Xmlio.checkXPath(doc, basexpath)) {
				scrJrefCrdnav = Xmlio.extractStringAttrUclc(doc, basexpath, itemtag, "Si", "sref", "scrJrefCrdnav", mask, SCRJREFCRDNAV);

				return true;
			};

			return false;
		};

		public HashSet<Integer> comm(
					StatShr comp
				) {
			HashSet<Integer> items = new HashSet<Integer>();

			if (scrJrefCrdnav.equals(comp.scrJrefCrdnav)) items.add(SCRJREFCRDNAV);

			return(items);
		};

		public HashSet<Integer> diff(
					StatShr comp
				) {
			HashSet<Integer> commitems;
			HashSet<Integer> diffitems;

			commitems = comm(comp);

			diffitems = new HashSet<Integer>(Arrays.asList(SCRJREFCRDNAV));
			for (Integer ci: commitems) diffitems.remove(ci);

			return(diffitems);
		};

	};

	/**
		* DpchEngData (full: DpchEngSessMsdcData)
		*/
	public class DpchEngData extends DpchEngMsdc {

		public static final int SCRJREF = 1;
		public static final int FEEDFENSSEC = 2;
		public static final int STATSHR = 3;

		public DpchEngData() {
			super(VecMsdcVDpch.DPCHENGSESSMSDCDATA);

			feedFEnsSec = new Feed("FeedFEnsSec");
			statshr = new StatShr("");
		};

		public Feed feedFEnsSec;
		public StatShr statshr;

		public String getSrefsMask() {
			ArrayList<String> ss = new ArrayList<String>();

			if (has(SCRJREF)) ss.add("scrJref");
			if (has(FEEDFENSSEC)) ss.add("feedFEnsSec");
			if (has(STATSHR)) ss.add("statshr");

			return StrMod.vectorToString(ss, ';');
		};

		public void readXML(
					Document doc
					, String basexpath
					, boolean addbasetag
				) {

			clear();

			if (addbasetag) basexpath = Xmlio.checkUclcXPaths(doc, basexpath, "DpchEngSessMsdcData");

			if (Xmlio.checkXPath(doc, basexpath)) {
				scrJref = Xmlio.extractStringUclc(doc, basexpath, "scrJref", "", mask, SCRJREF);
				if (feedFEnsSec.readXML(doc, basexpath, true)) add(FEEDFENSSEC);
				if (statshr.readXML(doc, basexpath, true)) add(STATSHR);
			} else {
				scrJref = "";
				feedFEnsSec = new Feed("FeedFEnsSec");
				statshr = new StatShr("");
			};
		};

	};

};

